main/lsp/notifications/initialized.cc

```diff
@@ -1,37 +1,43 @@
 #include "main/lsp/notifications/initialized.h"
+#include "common/kvstore/KeyValueStore.h"
 #include "main/lsp/LSPIndexer.h"
 
 using namespace std;
 
 namespace sorbet::realmain::lsp {
 
 InitializedTask::InitializedTask(LSPConfiguration &config)
-    : LSPDangerousTypecheckerTask(config, LSPMethod::Initialized), mutableConfig(config), indexer(nullptr){};
+    : LSPTask(config, LSPMethod::Initialized), mutableConfig(config), gs{}, kvstore{} {};
 
 void InitializedTask::preprocess(LSPPreprocessor &preprocessor) {
     mutableConfig.markInitialized();
+    this->preprocessor = &preprocessor;
 }
 
 void InitializedTask::index(LSPIndexer &indexer) {
-    // Hacky: We need to use the indexer, but with the WorkerPool from runSpecial. This + SorbetWorkspaceEdit are the
-    // only tasks to have this special requirement.
-    this->indexer = &indexer;
-}
+    // We need to pause during indexing so that nothing else will try to give work to the index thread while its state
+    // is being initialized in the typechecker.
+    preprocessor->pause();
 
-void InitializedTask::runSpecial(LSPTypechecker &typechecker, WorkerPool &workers) {
-    ENFORCE(this->indexer != nullptr);
-    indexer->initialize(updates, workers);
-    typechecker.initialize(move(updates), workers);
-    // TODO: Make asynchronous.
-    complete.Notify();
+    indexer.transferInitializeState(*this);
 }
 
-void InitializedTask::schedulerWaitUntilReady() {
-    complete.WaitForNotification();
+void InitializedTask::run(LSPTypecheckerInterface &typechecker) {
+    ENFORCE(this->gs != nullptr);
+    typechecker.initialize(*this, std::move(this->gs), std::move(this->kvstore));
+    typechecker.resumeTaskQueue(*this);
```

Comment:
This can just be `preprocessor->resume()`, right?  Then you don't need `LSPTypecheckerInterface::resumeTaskQueue` et al.

Reply:
Oh, do we have access to the preprocessor in tasks?

Reply:
We have access to `preprocessor` in the `preprocess` method, but only access to the typechecker in `run` :(